std: `compilesettings` show CLI compile/link flags

saem opened the pull request:
## Summary

`compilersettings`  now includes compiler and linker options that are
set via the command-line  `--passc/passl`  flags.

## Details

Prior to this change only pragma based compiler and linker options were
shown, via  `{.passC: "...".}`  or  `{.passL:"...".}` . Now  `vmops` ,
which implements the compile time querying, uses compiler and linker
option construction procedures from the  `options`  module to fulfill
the query.

The compiler and linker options procedures in  `options`  are newly
introduced, taking over logic previously carried out in the  `extccomp` 
module, which orchestrates backend C compiler and linker operations. As
part of this change  `extccomp`  has been updated to also use these
procedures so there is one way and place where these options are
constructed. Additionally, the   `tcompilesetting`  test has been
updated to ensure these continue to work.

Note: the  `compilesettings`  query for compile options does _not_
include module specific compile options, set via 
`{.localPassc: "...".}`  pragma. This could be implemented, but that
would likely require further changes, such as introducing a new query to
get all file compiler options vs file specific.

saem committed:
std: `compilesettings` show CLI compile/link flags

Summary
=======

`compilersettings` now includes compiler and linker options that are set
via the command-line `--passc/passl` flags.

Details
=====

Prior to this change only pragma based compiler and linker options were
shown, via `{.passC: "...".}` or `{.passL:"...".}`. Now `vmops`, which
implements the compile time querying, uses compiler and link option
construction procedures from `extccomp` module, which orchestrates C
compilation and linking, to fulfill the query.

As part of this change the `tcompilesetting` test has been updated to
ensure these continue to work.

Note: the `compilesettings` query for compile options does _not_ include
module specific compile options, set via `{.localPassc: "...".}` pragma.
This could be implemented, but that would likely require further
changes, like introducing a new query to get all file compiler options
vs file specific.

zerbina reviewed (commented):
Regarding the impact:
1. in addition to the pragma-specified options, `compilerOptions` previously also contained the `<ccompiler>.options.always` options (it still does)
2. the `compileOptions` query result now includes the `<ccompiler>.options.speed|debug|size` options
3. the `linkOptions` query result now include the linked C libraries

Number 2 is fine, I think, but number 3 I'm not so sure about, given that the libraries to link against aren't really options, in that sense.

As an alternative, `vmops` could append the options itself, like so:
```nim
# compile options
result = conf.compileOptions
for option in conf.compileOptionsCmd:
  if find(result, option, 0) < 0:
    result.add " " & option

# link options
result = conf.linkOptions & " " & conf.linkOptionsCmd.join(" ") 
```
This would also remove the `extccomp` dependency, at the cost of a small duplication of logic. It's only a soft suggestion, however, so feel free to ignore.

saem commented:
Hmm, maybe the query should be in `options` and only cover combining pragma and cli options, as the split storage is a bit more of a `ConfigRef` detail.

---

Separately, wrt the failing tests I guess clang is pickier about undefined symbols, I need to find another "no-op" like option.

saem committed:
refactor compile and link options retreival

`options` now does the combining for cli and pragma compile and link
options. This makes some sense as the split storage is a
`options.`ConfigRef` choice.

This also allows dropping the `extccomp` dependency in `vmops`, so a
win?

saem committed:
fix for macos and windows compiler tool chains

my silly linker options are not appreciated on those platforms, it's
sufficient to test this on linux only as there is nothing really platfrom
specific about this code.

zerbina commented on compiler/front/options.nim:
```diff
@@ -393,6 +393,16 @@ passStrTableField symbols
 passStrTableField arcToExpand
 
+proc getCompileOptionsStr*(conf: ConfigRef): string =
+  result = conf.compileOptions
```
```suggestion
  ## Returns the combination of the current C compile options and the
  ## global `--passC` options (combined in that exact order).
  ##
  ## Global `--passC` options already present in the current C compile
  ## options are *not* include again.
  result = conf.compileOptions
```

saem replied:
Thank you for suggesting these docs, I totally blanked on the writing docs for public definitions.

zerbina commented on compiler/front/options.nim:
```diff
@@ -393,6 +393,16 @@ passStrTableField symbols
+      result.add option
+
+proc getLinkOptionsStr*(conf: ConfigRef): string =
+  conf.linkOptions & " " & conf.linkOptionsCmd.join(" ")
```
```suggestion
  ## Returns the combination of the current C linker options and the global
  ## `--passL` options (combined in that exact order).
  conf.linkOptions & " " & conf.linkOptionsCmd.join(" ")
```

zerbina commented on tests/vm/tcompilesetting.nim:
```diff
@@ -1,9 +1,9 @@
 discard """
-cmd: "nim c --nimcache:build/myNimCache --nimblePath:myNimblePath $file"
+matrix: "--nimcache:build/myNimCache --nimblePath:myNimblePath --passc:'-fmax-errors=4' --passl:'-u dummysymboldoesnotexist'"
```
```suggestion
matrix: "--nimcache:build/myNimCache --nimblePath:myNimblePath --passc:'-fmax-errors=4' --passl:'-Xlinker --defsym -Xlinker dummysymboldoesnotexist=0x0'"
```
I cannot test it with the XCode toolchain, but I think that this should work.

saem replied:
LooooooL, no idea why I thought guarding the compiler query would change the outcome of the test considering the cli params. I'm a bit distracted today. 😮‍💨 

I'll get things building again on my mac (so slow) and see if this works there.

zerbina reviewed (approved):
Nice, I think having the combination logic in `options` is indeed better. I've left two suggestion regarding documentation of the new procedures.

Looking for a safe replacement of `-u`, I found `--defsym` (create a global symbol storing the provided address), which might work with the linker from the Xcode toolchain (it does with the MinGW `ld`, at least).

saem committed:
add doc comments as provided by zerbina

saem committed:
use `--defsym` linker option as a no-op

found and suggested by zerbina, this should effectively be a no-op and
work across linkers.

saem committed:
move linker options testing into a separate test

only tests on linux, but that should provide at least some coverage

saem commented:
I'm restricting linker flag testing to linux only for now, this is turning out to be a bear. 😬

saem committed:
some quoting causing windows issues?

saem commented:
/merge

saem commented:
/merge

huh, not sure what's with the kill 9, in that last run trying again.
